modules/fileicon/material.go

```
@@ -0,0 +1,150 @@
+
+func DefaultMaterialIconProvider() *MaterialIconProvider {
+	return &materialIconProvider
+}
+
+func (m *MaterialIconProvider) loadData() {
+	buf, err := options.AssetFS().ReadFile("fileicon/material-icon-rules.json")
+	if err != nil {
+		log.Error("Failed to read material icon rules: %v", err)
+		return
+	}
+	err = json.Unmarshal(buf, &m.rules)
+	if err != nil {
+		log.Error("Failed to unmarshal material icon rules: %v", err)
+		return
+	}
+
+	buf, err = options.AssetFS().ReadFile("fileicon/material-icon-svgs.json")
+	if err != nil {
+		log.Error("Failed to read material icon rules: %v", err)
+		return
+	}
+	err = json.Unmarshal(buf, &m.svgs)
+	if err != nil {
+		log.Error("Failed to unmarshal material icon rules: %v", err)
+		return
+	}
+	log.Debug("Loaded material icon rules and SVG images")
+}
+
+func (m *MaterialIconProvider) renderFileIconSVG(ctx reqctx.RequestContext, name, svg string) template.HTML {
+	data := ctx.GetData()
+	renderedSVGs, _ := data["_RenderedSVGs"].(map[string]bool)
+	if renderedSVGs == nil {
+		renderedSVGs = make(map[string]bool)
+		data["_RenderedSVGs"] = renderedSVGs
+	}
+	// This part is a bit hacky, but it works really well. It should be safe to do so because all SVG icons are generated by us.
+	// Will try to refactor this in the future.
+	if !strings.HasPrefix(svg, "<svg") {
+		panic("Invalid SVG icon")
+	}
+	svgID := "svg-mfi-" + name
+	svgCommonAttrs := `class="svg fileicon" width="16" height="16" aria-hidden="true"`
+	posOuterBefore := strings.IndexByte(svg, '>')
+	if renderedSVGs[svgID] && posOuterBefore != -1 {
+		return template.HTML(`<svg ` + svgCommonAttrs + `><use xlink:href="#` + svgID + `"></use></svg>`)
+	}
+	svg = `<svg id="` + svgID + `" ` + svgCommonAttrs + svg[4:]
+	renderedSVGs[svgID] = true
+	return template.HTML(svg)
+}
+
+func (m *MaterialIconProvider) FileIcon(ctx reqctx.RequestContext, entry *git.TreeEntry) template.HTML {
+	m.once.Do(m.loadData)
+
+	if m.rules == nil {
```

Why do we ignore errors during loading? Then we need to do this check

---

Because this is the simplest approach IMO, and nothing bad.